sdlib/d/gc/arena.d

```diff
@@ -751,3 +813,107 @@ unittest resizeLarge {
 	// Now full again:
 	assert(pd9.extent.hpd.full);
 }
+
+unittest resizeLargeGrow {
+	import d.gc.arena;
+	shared Arena arena;
+
+	auto base = &arena.base;
+	scope(exit) arena.base.clear();
+
+	import d.gc.emap;
+	static shared ExtentMap emap;
+	emap.tree.base = base;
+
+	import d.gc.region;
+	shared RegionAllocator regionAllocator;
+	regionAllocator.base = base;
+
+	arena.regionAllocator = &regionAllocator;
+
+	// Allocation 0: 35 pages:
+	auto ptr0 = arena.allocLarge(&emap, 35 * PageSize);
+	assert(ptr0 !is null);
+	auto pd0 = emap.lookup(ptr0);
+	assert(pd0.extent.address is ptr0);
+	assert(pd0.extent.size == 35 * PageSize);
+
+	// Allocation 1: 64 pages:
+	auto ptr1 = arena.allocLarge(&emap, 64 * PageSize);
+	assert(ptr1 !is null);
+	auto pd1 = emap.lookup(ptr1);
+	assert(pd1.extent.hpd == pd0.extent.hpd);
+	assert(pd1.extent.address is ptr1);
+	assert(pd1.extent.size == 64 * PageSize);
+
+	// Allocation 2: 128 pages:
+	auto ptr2 = arena.allocLarge(&emap, 128 * PageSize);
+	assert(ptr2 !is null);
+	auto pd2 = emap.lookup(ptr2);
+	assert(pd2.extent.hpd == pd1.extent.hpd);
+	assert(pd2.extent.address is ptr2);
+	assert(pd2.extent.size == 128 * PageSize);
+
+	// Try to grow allocation 0, but cannot, there is no space to grow:
+	assert(!arena.resizeLarge(&emap, pd0.extent, 36 * PageSize));
+
+	// Try to grow allocation 0 by zero; this will succeed:
+	assert(arena.resizeLarge(&emap, pd0.extent, 35 * PageSize));
+
+	// Try to grow allocation 2, but cannot, asked for one too many pages:
+	assert(!arena.resizeLarge(&emap, pd2.extent, 414 * PageSize));
+
+	// Grow allocation 2 successfully by 413 pages:
+	assert(arena.resizeLarge(&emap, pd2.extent, 413 * PageSize));
+	assert(pd2.extent.address is ptr2);
+	assert(pd2.extent.size == 413 * PageSize);
+
+	// Free allocation 1:
+	arena.free(&emap, pd1, ptr1);
+
+	// Grow allocation 0:
+	assert(arena.resizeLarge(&emap, pd0.extent, 44 * PageSize));
+	assert(pd0.extent.address is ptr0);
+	assert(pd0.extent.size == 44 * PageSize);
+
+	// Confirm that the extent correctly grew and remapped:
+	for (auto p = pd0.extent.address; p < pd0.extent.address + pd0.extent.size;
+	     p += PageSize) {
+		auto pd0probe = emap.lookup(p);
+		assert(pd0probe.extent == pd0.extent);
+	}
```

Comment:
You have not checked this unless you check that the extent is not mapped after its end.

Reply:
You also want to check that the pd's index is correct (right now I suspect it won't be).

Reply:
Sorry I missed the check past the end. The indices are still probably screwed up.